File: server/services/awsService.ts
import { S3Client, PutObjectCommand, DeleteObjectCommand, ListObjectsV2Command } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { GetObjectCommand } from '@aws-sdk/client-s3';
import multer from 'multer';
import multerS3 from 'multer-s3';

// Configure AWS SDK v3
const s3Client = new S3Client({
  region: process.env.AWS_REGION || 'ap-south-1',
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!
  }
});

class AWSService {
  private bucketName: string;

  constructor() {
    this.bucketName = process.env.AWS_BUCKET_NAME || 'lexora-assets';
  }

  // Create multer upload middleware for S3
  createUploadMiddleware(folder: string = 'uploads') {
    return multer({
      storage: multerS3({
        s3: s3Client as any,
        bucket: this.bucketName,
        metadata: (req, file, cb) => {
          cb(null, { fieldName: file.fieldname });
        },
        key: (req, file, cb) => {
          const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
          const extension = file.originalname.split('.').pop();
          cb(null, `${folder}/${uniqueSuffix}.${extension}`);
        }
      }),
      limits: {
        fileSize: 50 * 1024 * 1024 // 50MB limit
      },
      fileFilter: (req, file, cb) => {
        // Allow images, videos, and audio files
        const allowedMimes = [
          'image/jpeg',
          'image/png',
          'image/gif',
          'video/mp4',
          'video/avi',
          'video/mov',
          'audio/mp3',
          'audio/wav',
          'audio/mpeg'
        ];
        
        if (allowedMimes.includes(file.mimetype)) {
          cb(null, true);
        } else {
          cb(new Error('Invalid file type'));
        }
      }
    });
  }

  // Upload file directly to S3
  async uploadFile(file: Buffer, fileName: string, mimeType: string, folder: string = 'uploads'): Promise<string> {
    try {
      const command = new PutObjectCommand({
        Bucket: this.bucketName,
        Key: `${folder}/${fileName}`,
        Body: file,
        ContentType: mimeType,
        ACL: 'public-read'
      });

      await s3Client.send(command);
      return `https://${this.bucketName}.s3.${process.env.AWS_REGION || 'ap-south-1'}.amazonaws.com/${folder}/${fileName}`;
    } catch (error) {
      console.error('S3 Upload Error:', error);
      throw new Error('Failed to upload file to S3');
    }
  }

  // Delete file from S3
  async deleteFile(fileKey: string): Promise<void> {
    try {
      const command = new DeleteObjectCommand({
        Bucket: this.bucketName,
        Key: fileKey
      });

      await s3Client.send(command);
    } catch (error) {
      console.error('S3 Delete Error:', error);
      throw new Error('Failed to delete file from S3');
    }
  }

  // Get signed URL for private files
  async getSignedUrl(fileKey: string, expiresIn: number = 3600): Promise<string> {
    const command = new GetObjectCommand({
      Bucket: this.bucketName,
      Key: fileKey
    });

    return await getSignedUrl(s3Client, command, { expiresIn });
  }

  // List files in a folder
  async listFiles(folder: string): Promise<any[]> {
    try {
      const command = new ListObjectsV2Command({
        Bucket: this.bucketName,
        Prefix: folder + '/'
      });

      const result = await s3Client.send(command);
      return result.Contents || [];
    } catch (error) {
      console.error('S3 List Error:', error);
      throw new Error('Failed to list files from S3');
    }
  }
}

export default new AWSService();

File: server/Wav2Lip/app/aws_service.py
import boto3
import os
from datetime import datetime
from dotenv import load_dotenv

# Load environment variables
load_dotenv(os.path.join(os.path.dirname(__file__), '..', '..', '.env'))

class AWSService:
    def __init__(self):
        self.s3_client = boto3.client(
            's3',
            region_name=os.getenv('AWS_REGION', 'ap-south-1'),
            aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
            aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY')
        )
        self.bucket_name = os.getenv('AWS_BUCKET_NAME', 'lexora-assets')
    
    def upload_video(self, video_path, session_id):
        """
        Upload generated video to S3 and return the public URL
        """
        try:
            # Generate a unique filename
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            s3_key = f"generated-videos/video_{session_id}_{timestamp}.mp4"
            
            # Upload to S3
            with open(video_path, 'rb') as video_file:
                self.s3_client.upload_fileobj(
                    video_file,
                    self.bucket_name,
                    s3_key,
                    ExtraArgs={
                        'ContentType': 'video/mp4',
                        'ACL': 'public-read'
                    }
                )
            
            # Return the public URL
            s3_url = f"https://{self.bucket_name}.s3.{os.getenv('AWS_REGION', 'ap-south-1')}.amazonaws.com/{s3_key}"
            print(f"Video uploaded successfully to S3: {s3_url}")
            return s3_url
            
        except Exception as e:
            print(f"Error uploading video to S3: {e}")
            return None
    
    def upload_audio(self, audio_path, session_id):
        """
        Upload generated audio to S3 and return the public URL
        """
        try:
            # Generate a unique filename
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            s3_key = f"generated-audios/audio_{session_id}_{timestamp}.mp3"
            
            # Upload to S3
            with open(audio_path, 'rb') as audio_file:
                self.s3_client.upload_fileobj(
                    audio_file,
                    self.bucket_name,
                    s3_key,
                    ExtraArgs={
                        'ContentType': 'audio/mpeg',
                        'ACL': 'public-read'
                    }
                )
            
            # Return the public URL
            s3_url = f"https://{self.bucket_name}.s3.{os.getenv('AWS_REGION', 'ap-south-1')}.amazonaws.com/{s3_key}"
            print(f"Audio uploaded successfully to S3: {s3_url}")
            return s3_url
            
        except Exception as e:
            print(f"Error uploading audio to S3: {e}")
            return None

# Create a singleton instance
aws_service = AWSService()
